### PR TITLE
279 wrong canvas size

### LINKED ISSUE #279: Wrong canvas size when the output is not perfectly squared
When I create a graphical Kernel with an square output everything is correct, but if the output is rectangular then either the height or the width will be wrong. For example, with this code:

```
const gpu = new GPU();
const WIDTH = 600;
const HEIGHT = 400;

var matrix;
var body;

const initMatrix = gpu.createKernel(function(value) {  
  return value;
})
.setOutput([WIDTH, HEIGHT]);

const render = gpu.createKernel(function(matrix) {
  var i = matrix[this.thread.y][this.thread.x];
  this.color(i, i, i, 1);
})
.setOutput([WIDTH, HEIGHT])
.setGraphical(true);

const canvas = render.getCanvas();

$(document).ready(function() {
  matrix = initMatrix(0.5);
  render(matrix);
  body = $('body')[0];
  body.appendChild(canvas);
  console.log(canvas);
});
```

I get a 600 x 490 canvas instead of 600x400: [Capture](https://i.gyazo.com/3d55bc2d4545a7fb8a2c3f9dc7f5ec6e.png)

### PR DESCRIPTION
**Pull request by nickhingston**
performance improvement (dont transfer more mem than necessary) and fix #279 - wrong output canvas size

### CODE REVIEW

File: test/issues/279-wrong-canvas-size.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	}
+	
+	QUnit.test('Issue #314 Large array addressing - auto', () => {
+		
+		var canvas = buildKernel().getCanvas();
+		QUnit.assert.equal(canvas.width, WIDTH);
+		QUnit.assert.equal(canvas.height, HEIGHT);
+	});
+
+	// QUnit.test('Issue #314 Large array addressing - gpu', () => {
```
Comment (robertleeplummerjr) [inline test/issues/279-wrong-canvas-size.js]: Was this meant to be left in?

File: test/issues/96-param-names.js [hunk 2]
```diff
@@ -26,32 +26,40 @@
     }, function (a, b) {
```
Comment (robertleeplummerjr) [inline test/issues/96-param-names.js]: Can you get these variables to match in name with those tests below or vise versa?

File: src/backend/web-gl/kernel.js [hunk 3]
```diff
@@ -700,8 +700,21 @@ module.exports = class WebGLKernel extends KernelBase {
...
-					const valuesFlat = new Float32Array(length);
-					utils.flattenTo(value, valuesFlat);
+					let valuesFlat
+
+					if (utils.isArray(value[0])) {
+						// not already flat
+						valuesFlat = new Float32Array(length);
+						utils.flattenTo(value, valuesFlat);
+					} else if (value.constructor != Float32Array) {
+						// TODO: would be great if we could not have to create Float32Array buffers
```
Comment (robertleeplummerjr) [inline src/backend/web-gl/kernel.js]: Should we open an issue?

File: test/issues/279-wrong-canvas-size.js (hunk 1 above)
Comment (nickhingston) [reply test/issues/279-wrong-canvas-size.js]: yes, and the title should not be 'Issue #314 Large array addressing'

will fix

File: src/backend/web-gl/kernel.js (hunk 3 above)
Comment (nickhingston) [reply src/backend/web-gl/kernel.js]: could do, was what we were talking about yesterday!

File: test/issues/96-param-names.js (hunk 2 above)
Comment (nickhingston) [reply test/issues/96-param-names.js]: not sure what you mean?

File: src/backend/web-gl/kernel.js (hunk 3 above)
Comment (robertleeplummerjr) [reply src/backend/web-gl/kernel.js]: https://github.com/gpujs/gpu.js/issues/346

ty @nickhingston!!!!

**Comment by robertleeplummerjr** [comment]
https://github.com/gpujs/gpu.js/releases/tag/1.5.2

**Pull request merged**